service/dynamodb/endpoints.go

```diff
@@ -508,33 +506,32 @@ func (m *resolveEndpointV2Middleware) HandleSerialize(ctx context.Context, in mi
 		return out, metadata, fmt.Errorf("expected endpoint resolver to not be nil")
 	}
 
-	params := bindEndpointParams(in.Parameters, m.options)
-	resolvedEndpoint, err := m.options.EndpointResolverV2.ResolveEndpoint(ctx, *params)
+	params := bindEndpointParams(getOperationInput(ctx), m.options)
+	endpt, err := m.options.EndpointResolverV2.ResolveEndpoint(ctx, *params)
 	if err != nil {
 		return out, metadata, fmt.Errorf("failed to resolve service endpoint, %w", err)
 	}
 
-	req.URL = &resolvedEndpoint.URI
-	for k := range resolvedEndpoint.Headers {
-		req.Header.Set(k, resolvedEndpoint.Headers.Get(k))
+	if endpt.URI.RawPath == "" && req.URL.RawPath != "" {
```

Comment:
im guessing this logic was pulled from `OperationSerializer`? should we maybe put that into its own function and document it or just add a comment so it doesnt feel out of place?
and if thats true, was this logic removed from `OperationSerializer` so its not duplicated (since i dont see a change in `serializers.go`?

Reply:
> im guessing this logic was pulled from OperationSerializer? should we maybe put that into its own function and document it or just add a comment so it doesnt feel out of place?

We _could_ but then it's part of the public API for something that is very much internal. Defaulting to "don't export unless necessary" here, duplication aside.

> was this logic removed from OperationSerializer so its not duplicated (since i dont see a change in serializers.go?

It wasn't, I will check to see if it's safe to remove this. If we just hard set the path in serialization before endpoints2 then I'd say it should be removed (as it exists right now after this change it basically does nothing because the path at that point is empty - **unless** caller middleware did something to it). This is a subtle thing behaviorally but if we've _always_ constructed the path in serde and joined it to anything existing then that's a behavioral quirk we should leave be. On the flip side, if it wasn't there before EP2 we should drop it.

Reply:
> We could but then it's part of the public API for something that is very much internal. Defaulting to "don't export unless necessary" here, duplication aside.

oh i more meant an unexported function. pulling it out into its own function was more of a code organization comment. but nvm, its fine as is.

> It wasn't, I will check to see if it's safe to remove this
sounds good

> if we've always constructed the path in serde and joined it to anything existing then that's a behavioral quirk we should leave be.
is there a behavioral difference though between constructing the path in serialize instead of finalize? are there specific customer hooks between serialize and finalize where moving this logic from serialize to finalize makes a difference?

Reply:
It looks like we've always "appended" to any existing path at the serialize step even before endpoint resolution, so that's going to have to remain.

e.g. https://github.com/aws/aws-sdk-go-v2/blob/v1.17.7/service/dynamodb/serializers.go#L592

Essentially the change here is that instead of resolving endpoint and then serializing we're now doing that in reverse. So what was once a one-step `join(existing url that happens to have resolved endpoint, serde)` is now happening in two parts:
1. `join(some caller injected url, serde)`
2. `join(resolved endpoint, result of step 1)`